crates/client/block_import/src/pre_validate.rs

```diff
@@ -0,0 +1,336 @@
+        converted_classes,
+    })
+}
+
+fn block_commitments(
+    block: &UnverifiedFullBlock,
+    validation: &Validation,
+) -> Result<ValidatedCommitments, BlockImportError> {
+    let (mut receipt_c, mut state_diff_c, mut transaction_c, mut event_c) = Default::default();
+    [
+        Box::new(|| {
+            receipt_c = receipt_commitment(block, validation)?;
+            Ok(())
+        }) as Box<dyn FnOnce() -> Result<(), BlockImportError> + Send>,
+        Box::new(|| {
+            state_diff_c = state_diff_commitment(block, validation)?;
+            Ok(())
+        }),
+        Box::new(|| {
+            transaction_c = transaction_commitment(block, validation)?;
+            Ok(())
+        }),
+        Box::new(|| {
+            event_c = event_commitment(block, validation)?;
+            Ok(())
+        }),
+    ]
+    .into_par_iter()
+    .map(|f| f())
+    .collect::<Result<(), _>>()?;
+
+    Ok(ValidatedCommitments {
+        transaction_count: block.transactions.len() as _,
+        transaction_commitment: transaction_c,
+        event_count: block.receipts.iter().map(|r| r.events().len()).sum::<usize>() as _,
+        event_commitment: event_c,
+        state_diff_length: block.state_diff.len() as _,
+        state_diff_commitment: state_diff_c,
+        receipt_commitment: receipt_c,
+    })
+}
+
+fn convert_classes(
+    validation: &Validation,
+    declared_classes: Vec<DeclaredClass>,
+) -> Result<Vec<ConvertedClass>, BlockImportError> {
+    declared_classes.into_par_iter().map(|class| class_conversion(validation, class)).collect()
+}
+
+fn class_conversion(_validation: &Validation, class: DeclaredClass) -> Result<ConvertedClass, BlockImportError> {
+    let DeclaredClass { class_hash, contract_class, compiled_class_hash } = class;
+
+    // TODO(class_hash): uncomment this when the class hashes are computed correctly accross the entire state
+    // let expected =
+    //     contract_class.class_hash().map_err(|e| BlockImportError::ComputeClassHashError(e.to_string()))?;
+    // if class_hash != expected {
+    // }
+
+    let compiled_class = contract_class
+        .compile()
+        .map_err(|e| BlockImportError::CompilationClassError { error: e.to_string(), class_hash })?;
+
+    let class_info = ClassInfo { contract_class, compiled_class_hash };
+
+    Ok(ConvertedClass { class_infos: (class_hash, class_info), class_compiled: (class_hash, compiled_class) })
+}
+
+fn transaction_hashes(
+    receipts: &[TransactionReceipt],
+    transactions: &[Transaction],
+    validation: &Validation,
+    block_n: Option<u64>,
+) -> Result<Vec<Felt>, BlockImportError> {
+    if receipts.len() != transactions.len() {
+        return Err(BlockImportError::TransactionEqualReceiptCount {
+            receipts: receipts.len(),
+            transactions: transactions.len(),
+        });
+    }
+
+    // mismatched block hash is allowed for blocks 1466..=2242 on mainnet
+    let is_special_trusted_case =
```

Comment:
that should be noted that an authority node cannot be run with the chain id `SN_MAIN` then or this will be triggered